include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -271,8 +300,84 @@ namespace RINGMesh
 
         GEO::AttributesManager attributes_manager_;
     };
+    ALIAS_2D_AND_3D( CartesianGridBase );
+
+    template < index_t DIMENSION >
+    class CartesianGrid final : public CartesianGridBase< DIMENSION >
+    {
+        friend class CartesianGridBuilder< DIMENSION >;
+    };
     ALIAS_2D_AND_3D( CartesianGrid );
 
+    template <>
+    class CartesianGrid< 3 > final : public CartesianGridBase< 3 >
+    {
+        friend class CartesianGridBuilder< 3 >;
+
+    public:
+        CartesianGrid( ivec3 nb_cells_in_each_direction,
+            ReferenceFrame3D vec_cartesian_axis )
+            : CartesianGridBase(
+                  nb_cells_in_each_direction, vec_cartesian_axis ),
+              grid_cage_()
+        {
+            for( auto i : range( 3 ) )
+            {
+                vec3 origin1 = cartesian_frame_.origin();
+                vec3 vector1 = cartesian_frame_[i];
+                grid_cage_.emplace_back( Geometry::Plane{ vector1, origin1 } );
+                vec3 origin2 = cartesian_frame_.origin()
+                               + cartesian_frame_[0] * nb_cells_axis( 0 )
+                               + cartesian_frame_[1] * nb_cells_axis( 1 )
+                               + cartesian_frame_[2] * nb_cells_axis( 2 );
+                vec3 vector2 = cartesian_frame_[i];
+                grid_cage_.emplace_back( Geometry::Plane{ vector2, origin2 } );
+            }
+        }
+
+        const std::vector< Geometry::Plane >& grid_cage() const
+        {
+            return grid_cage_;
+        }
+
+    private:
+        std::vector< Geometry::Plane > grid_cage_;
+    };
+
+    template <>
+    class CartesianGrid< 2 > final : public CartesianGridBase< 2 >
+    {
+        friend class CartesianGridBuilder< 2 >;
+
+    public:
+        CartesianGrid( ivec2 nb_cells_in_each_direction,
+            ReferenceFrame2D vec_cartesian_axis )
+            : CartesianGridBase(
+                  nb_cells_in_each_direction, vec_cartesian_axis )
+        {
+        }
+
+        const std::vector< Geometry::Segment2D > grid_cage() const
+		{
+        	std::vector< Geometry::Segment2D > cage;
+        	cage.resize(4);
+        	Geometry::Segment2D seg1{ vec2{cartesian_frame_.origin()},
+        				vec2{cartesian_frame_.origin()+cartesian_frame_[0]} };
+			cage.emplace_back( seg1 );
+			Geometry::Segment2D seg2{ vec2{cartesian_frame_.origin()},
+        				vec2{cartesian_frame_.origin()+cartesian_frame_[1]} };
+			cage.emplace_back( seg2 );
+			Geometry::Segment2D seg3{ vec2{cartesian_frame_.origin()+cartesian_frame_[0]},
+				vec2{cartesian_frame_.origin()+cartesian_frame_[0]+cartesian_frame_[1]} };
+			cage.emplace_back( seg3 );
+			Geometry::Segment2D seg4{ vec2{cartesian_frame_.origin()+cartesian_frame_[1]},
+				vec2{cartesian_frame_.origin()+cartesian_frame_[0]+cartesian_frame_[1]} };
+			cage.emplace_back( seg4 );
```

Comment:
Maybe you can find a way to factorize these 4 segments generation using a function

Reply:
I thought about it, but seeing how the vectors are created, I think it would just complicate things more than simplify them.

Reply:
OK